cruise-control/src/main/java/com/linkedin/kafka/cruisecontrol/executor/ExecutionUtils.java

```diff
@@ -30,10 +42,119 @@
   public static final String GAUGE_EXECUTION_STOPPED_BY_USER = EXECUTION_STOPPED + "-by-user";
   public static final String GAUGE_EXECUTION_STARTED_IN_KAFKA_ASSIGNER_MODE = EXECUTION_STARTED + "-" + KAFKA_ASSIGNER_MODE;
   public static final String GAUGE_EXECUTION_STARTED_IN_NON_KAFKA_ASSIGNER_MODE = EXECUTION_STARTED + "-non-" + KAFKA_ASSIGNER_MODE;
+  public static final long EXECUTION_HISTORY_SCANNER_PERIOD_SECONDS = 5;
+  public static final long EXECUTION_HISTORY_SCANNER_INITIAL_DELAY_SECONDS = 0;
+  public static final int ADDITIVE_INCREASE_PARAM = 1;
+  public static final int MULTIPLICATIVE_DECREASE_PARAM = 2;
+  static final Map<String, Double> CONCURRENCY_ADJUSTER_LIMIT_BY_METRIC_NAME = new HashMap<>(4);
 
 
   private ExecutionUtils() { }
 
+  /**
+   * Initialize the concurrency adjuster limits.
+   *
+   * @param config The configurations for Cruise Control.
+   */
+  static void init(KafkaCruiseControlConfig config) {
+    CONCURRENCY_ADJUSTER_LIMIT_BY_METRIC_NAME.put(BROKER_LOG_FLUSH_TIME_MS_999TH.name(),
+                                                  config.getDouble(ExecutorConfig.CONCURRENCY_ADJUSTER_LIMIT_LOG_FLUSH_TIME_MS_CONFIG));
+    CONCURRENCY_ADJUSTER_LIMIT_BY_METRIC_NAME.put(BROKER_FOLLOWER_FETCH_LOCAL_TIME_MS_999TH.name(),
+                                                  config.getDouble(ExecutorConfig.CONCURRENCY_ADJUSTER_LIMIT_FOLLOWER_FETCH_LOCAL_TIME_MS_CONFIG));
+    CONCURRENCY_ADJUSTER_LIMIT_BY_METRIC_NAME.put(BROKER_PRODUCE_LOCAL_TIME_MS_999TH.name(),
+                                                  config.getDouble(ExecutorConfig.CONCURRENCY_ADJUSTER_LIMIT_PRODUCE_LOCAL_TIME_MS_CONFIG));
+    CONCURRENCY_ADJUSTER_LIMIT_BY_METRIC_NAME.put(BROKER_CONSUMER_FETCH_LOCAL_TIME_MS_999TH.name(),
+                                                  config.getDouble(ExecutorConfig.CONCURRENCY_ADJUSTER_LIMIT_CONSUMER_FETCH_LOCAL_TIME_MS_CONFIG));
```

Comment:
I noticed that the metrics used by the adjuster are all broker-side metrics related to consumption and production. It may not reflect problems related to delayed leadership changes. For example, there can be a long queue of LeaderAndIsrRequest or UpdateMetadataRequest causing some partitions to become leaderless while the fetch/produce local time metrics still look good because there is less I/O if partitions are leaderless.

Have you considered also looking at the controller metrics as well? The ones I am thinking about are:
- Controller to per broker request queue time: https://github.com/linkedin/kafka/blob/2.3-li-1/core/src/main/scala/kafka/controller/ControllerChannelManager.scala#L184
- Controller to per broker request queue size: https://github.com/linkedin/kafka/blob/2.3-li-1/core/src/main/scala/kafka/controller/ControllerChannelManager.scala#L192

This patch is already a big step forward and I don't think we should address this comment in this patch but it is something worth thinking.

Reply:
Thanks for the feedback -- We can definitely look into enriching the set of tracked broker metrics! 

Overall, I feel that bursty requests might lead to an increase in the queue size, but the growing size would be acceptable as long as the per-request time spent on the queue does not increase beyond "some acceptable" limit. That being said, as a safety net, we can track `RequestQueueSize` with a reasonable limit (_updated the code to add this_).
Controller to per broker request queue time metric is interesting, but it breaks CC's metric reporting/sampling assumptions on 1-1 mapping between metrics and brokers -- i.e. all other broker metrics are reported per broker, however, in this case controller reports as many metrics as the number of brokers in the cluster.

Reply:
Some scenarios may not be reflected in `RequestQueueSize` on the broker side because controller is doing blocking send and receive for the controller request, i.e. it waits for the broker response before sending out the next request in queue. 

One example can be:
- Controller receives a lot of operations (e.g. partition reassignment, preferred leader election) at the same time, which results in a large number of requests enqueued in the controller broker request queue. Let's say there are 2001 requests needed to be sent out per broker.
- The broker is processing the controller request at a normal speed. There is no excessive load on the broker so broker side metrics look fine, including `RequestQueueSize` and `Request LocalTime` metrics. Let's say the average latency for processing each request is 50ms.
- Due to the natural of the blocking send and receive for controller request, the 2001th request in the controller broker queue can take 2000 * 50ms = 100s to reach the broker since the request is generated. If the controller broker request queue is imbalance across different brokers, it can easily lead to prolong leadership transition time for some partitions.

Reply:
The controller indeed has a blocking send on a separate thread for each broker -- i.e. https://github.com/linkedin/kafka/blob/2.3-li-1/core/src/main/scala/kafka/controller/ControllerChannelManager.scala#L274. 

In arguably the typical case with more data plane requests (e.g. produce/consume), a backlogged `requestQueue` on the broker side would cause the relevant per-broker controller thread to be blocked  (i.e. controller cannot send out more requests to that broker) for a longer period of time on the broker response w/o the `KIP-291` support.

I believe the scenario you exemplified can happen if the controller receives a burst of control plane requests (i.e. `LeaderAndIsr`, `StopReplica`, or `UpdateMetadata`). This is more likely to happen during a 
1. Demote operation, 
2. PLE, 
3. Replication factor change, or 
4. A leadership transfer phase of the other goal-based CC executions, which are configured with a high "leadership" movement concurrency (i.e. leadership movement concurrency also applies to replica reorder concurrency during the demote operation -- this generates `UpdateMetadata` requests) and/or low "execution progress check" interval. 

Except for the replication factor changes, which should now be addressed with the latest commit, changes introduced in this PR were already _not_ relevant to these operations. Also note that Concurrency Adjuster comes with a `concurrency.adjuster.max.partition.movements.per.broker` config, which puts a bound on the extent of concurrency increase. Hence, even if this becomes an issue, there are means to workaround it.

However, once we want to add support to cover auto-concurrency adjustment for the listed cases above, controller metrics will be more relevant. Hence, I created https://github.com/linkedin/cruise-control/issues/1292 to address those cases.